DQM/src/DQM/VisiblesFeatureProducer.cxx

```diff
@@ -0,0 +1,328 @@
+              // have vertical strips
+              if (hit.isOrientationX()) {
+                if (fabs(hit2.getYPos() - hit_y) > 0) {
+                  if (fabs(hit2.getYPos() - hit_y) < closest_point) {
+                    closest_point = fabs(hit2.getYPos() - hit_y);
+                  }
+                }
+              }
+              if (hit.isOrientationY()) {
+                if (fabs(hit2.getXPos() - hit_x) > 0) {
+                  if (fabs(hit2.getXPos() - hit_x) < closest_point) {
+                    closest_point = fabs(hit2.getXPos() - hit_x);
+                  }
+                }
+              }
+            }
+          }
+        }
+        if (closest_point > 50.) {
+          iso_hits += 1;
+          iso_energy += hit.getEnergy();
+        }
+      }
+    }
+
+    n_layers_hit = layers_hit.size();
+
+    if (summed_det > 0.) {
+      x_mean /= summed_det;
+      y_mean /= summed_det;
+      z_mean /= summed_det;
+      r_mean /= summed_det;
+
+      r_mean_from_photon_track /= summed_det;
+    }
+
+    for (const ldmx::HcalHit &hit : hcal_rec_hits) {
+      if (hit.getEnergy() > 0.) {
+        if (fabs(hit.getXPos()) > 1000 || fabs(hit.getYPos()) > 1000) {
+          continue;
+        }
+        ldmx::HcalID detID(hit.getID());
+        if (detID.getSection() == 0) {
+          x_std += hit.getEnergy() * (hit.getXPos() - x_mean) *
+                   (hit.getXPos() - x_mean);
+          y_std += hit.getEnergy() * (hit.getYPos() - y_mean) *
+                   (hit.getYPos() - y_mean);
+          z_std += hit.getEnergy() * (hit.getZPos() - z_mean) *
+                   (hit.getZPos() - z_mean);
+        }
+      }
+    }
+
+    if (summed_det > 0.) {
+      x_std = sqrt(x_std / summed_det);
+      y_std = sqrt(y_std / summed_det);
+      z_std = sqrt(z_std / summed_det);
+    }
+
+    // Fill histograms
+    histograms_.fill("layers_hit", n_layers_hit);
+    histograms_.fill("x_std", x_std);
+    histograms_.fill("y_std", y_std);
+    histograms_.fill("z_std", z_std);
+    histograms_.fill("x_mean", x_mean);
+    histograms_.fill("y_mean", y_mean);
+    histograms_.fill("r_mean", r_mean);
+    histograms_.fill("iso_hits", iso_hits);
+    histograms_.fill("iso_energy", iso_energy);
+    histograms_.fill("n_hits", n_readout_hits);
+    histograms_.fill("total_energy", hcal_energy);
+    histograms_.fill("photon_track", r_mean_from_photon_track);
+
+    bdt_features.push_back(n_layers_hit);
+    bdt_features.push_back(x_std);
+    bdt_features.push_back(y_std);
+    bdt_features.push_back(z_std);
+    bdt_features.push_back(x_mean);
+    bdt_features.push_back(y_mean);
+    bdt_features.push_back(r_mean);
+    bdt_features.push_back(iso_hits);
+    bdt_features.push_back(iso_energy);
+    bdt_features.push_back(n_readout_hits);
+    bdt_features.push_back(hcal_energy);
+    bdt_features.push_back(r_mean_from_photon_track);
+
+    if (training_) {
+      std::ofstream file(training_file_, std::ios::app);
```

Comment:
Why do you have a separate text CSV file for training?

I feel like we could put the bdt features into the event bus and then use `uproot` on the Python side to load these arrays into a table (probably pandas.Dataframe or pytorch.Tensor). This comment is not blocking this PR, I'm just using it as an opportunity to point out that you can add "simple" objects to the event bus and use them later.

Reply:
No worries, sounds like a good plan :)